### PR TITLE
[MLIR] Add support for frame pointers in MLIR

### PR DESCRIPTION
**Pull request by Radu2k**
Add support for frame pointers in MLIR.

### CODE REVIEW

File: mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp
Comment (zero9178) [inline mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp:2312]: Seems there's an editor artifact here :)

File: mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp
Comment (zero9178) [inline mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp:2331]: Emitting errors in parsing code should be done via the `emitError` method of `parser` as this utilizes MLIRs diagnostics infrastructure. Writing to `llvm::outs` or any other global stream in particularly is heavly discouraged.

File: mlir/lib/Target/LLVMIR/ModuleTranslation.cpp
Comment (zero9178) [inline mlir/lib/Target/LLVMIR/ModuleTranslation.cpp:899]: Could you spell out the `auto` here?

File: mlir/test/Target/LLVMIR/frame-pointer.mlir
Comment (zero9178) [inline mlir/test/Target/LLVMIR/frame-pointer.mlir:5]:
```suggestion
  // CHECK-LABEL: define void @frame_pointer_func() #[[ATTRS:\d+]]
  // CHECK: attributes #[[ATTRS]] = { "frame-pointer"="non-leaf" }
```
you can make the test more robust to future changes here by making sure that the attribute set corresponds to the attribute set of the function

**Review by zero9178 — COMMENTED** [review]
Thank you for working on this! I think this looks mostly good but left some comments in the code.

I only have one high level question: Is it necassary to define a custom syntax for frame pointers within `llvm.func` or could it just be printed as part the dictionary attribute of `llvm.func`? 

I personally believe that `llvm.func`s syntax has gained too much custom syntax for function attributes that are simply not required. With the growing number of LLVM function attributes being added and implemented I also fear that this could get out of hand. 
Lastly it would also be less code :)

Adding some other reviewers as well

File: mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp
Comment (Dinistro) [inline mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp:2328]: Nit: Consider splitting the computation and the check by moving the computation before the if. This allows you to reuse the result in case of a success.

**Review by Dinistro — COMMENTED** [review]
Thanks for this extension 😄 

The code is looking fine, modulo a formatting issue (I suggest to either enable auto-formatting or setting up a pre-commit hock with clang-tidy-diff).
I do agree with @zero9178 on the skepticism of adding this to the custom assembly format. Keeping it in the attribute dictionary seems to be enough, considering that this will not be used by all users of `LLVMFuncOp`.

**Comment by zero9178** [comment]
> Thanks for the valuable feedback!
> 
> Regarding @zero9178's proposal, it's important to note that this value is an enum, limited to specific values. There are three potential approaches for implementation:
> 
> 1. Use a Separate Enum Type: This is the existing approach. Continue using the enum to ensure only permissible values are used. Implementation across the code is managed through autogenerated functions.
> 2. Dictionary Array Attribute: Can we maintain this with the dictionary array attribute? Could you give an example of an attribute that uses an enumeration but works with the dictionary?
> 3. String Attribute: Would it be more appropriate to define this as a string attribute?

I believe there is a misunderstanding here. I think the use of an enum attribute to model the possible values of frame pointers is perfect :)

What I was suggesting is that the `llvm.func` op parser and printer shouldn't need to have special syntax for handling the frame pointer attribute. If at the call to `function_interface_impl::printFunctionAttributes` you simply don't add the frame pointer attribute name to the list it should be printed as part of the attribute dictionary that all non-specially handled attributes are placed in.

File: mlir/test/Target/LLVMIR/frame-pointer.mlir
Comment (Radu2k) [reply mlir/test/Target/LLVMIR/frame-pointer.mlir]: Small update on this topic, ATTRS cannot be defined in CHECK-LABEL and we have to use CHECK-SAME just below CHECK-LABEL, as well \d is not working but [0-9]+ works. Will update in next commit, thanks!

File: mlir/test/Target/LLVMIR/Import/frame-pointer.ll
Comment (Dinistro) [inline mlir/test/Target/LLVMIR/Import/frame-pointer.ll:5]: It seems that the EnumAttr is not capable of forcing the assembly format of it to be a string. I suspect that something like the `LinkageAttr`, see `LLVMAttrDefs.td` is required to get nice printing.
Even though I worked with this at some point, I cannot explain to you why this is necessary.

Alternatively, it might be possible to add an assembly format to the existing attr, but I'm not sure this works.

**Review by Dinistro — COMMENTED** [review]
Thanks for addressing the comments. I added a comment related to the printing of this new attr.

File: mlir/lib/Target/LLVMIR/ModuleImport.cpp
Comment (gysit) [inline mlir/lib/Target/LLVMIR/ModuleImport.cpp:1605]: Let's maybe update the comment above to something more generic not enumerating all the attributes (since this will be outdated once we add the next explicit attribute. What about putting something like this:

    // Skip all function attributes that have an explicit attribute in the 
    // LLVMFuncOp argument list and thus should not be stored in the 
    // passthrough attribute dictionary.

File: mlir/lib/Target/LLVMIR/ModuleTranslation.cpp
Comment (gysit) [inline mlir/lib/Target/LLVMIR/ModuleTranslation.cpp:899]:
```suggestion
  if (FramePointerKindAttr attr = func.getFramePointerAttr())
```
ultra nit: I think you can drop the namespace due to the using mlir::LLVM at the top of the file.

**Review by gysit — COMMENTED** [review]
I have only a few nit comments already things look good once the attribute prints nicely. I  believe @Dinistro's idea of introducing a dialect attribute should work.

File: mlir/lib/Target/LLVMIR/ModuleImport.cpp
Comment (Dinistro) [reply mlir/lib/Target/LLVMIR/ModuleImport.cpp:1605]: Note that this has changed with a recently landed commit, so maybe rebase first.

Comment (gysit) [reply mlir/lib/Target/LLVMIR/ModuleImport.cpp:1605]: Ah was not aware of that. Rebasing will resolve the comment then.

**Comment by Radu2k** [comment]
@zero9178 @Dinistro 

Tried to adjust the code so that it produces the appropriate formatting, but mimicking the **_LinkageAttr_** implementation gives  `#llvm< [mnemonic]"[FPvalue]">` when transforming the LLVM IR into MLIR, whereas lowing to LLVM IR from MLIR does not work. 

Based on my investigation, this is due to the fact that it is declared as an LLVM_Attr which extends AttrDef which also extends DialectAttr and the final attribute is treated as a Dialect attribute in **_AsmPrinter.cpp_** `printDialectAttribute` which prints it as `printDialectSymbol(os, "#", dialect.getNamespace(), attrName);`.

I have tried to change the attribute type and could not find a solution without adding some lines of code in _**LLVMDialect.cpp**_ (Going back to the initial issue).  Please let me know how we should treat this issue, and if there is any possibility to have this implemented without further complicating the dialect, will surely give it a go.

File: mlir/include/mlir/Dialect/LLVMIR/LLVMAttrDefs.td
Comment (Dinistro) [inline mlir/include/mlir/Dialect/LLVMIR/LLVMAttrDefs.td:56]:
```suggestion
  let assemblyFormat = "`<` $framePointerKind `>`";
```

Apparently, `<` `>` are required around the assembly format to make this attribute "simple enough for the pretty form".  https://github.com/llvm/llvm-project/blob/59e24eea3438ab2925bda025f52d2cce92195065/mlir/lib/IR/AsmPrinter.cpp#L2090-L2105

**Review by Dinistro — COMMENTED** [review]
Thanks for addressing the comment.
I think that I found the issue you were having. Once this is addressed, this PR will be good to go 🙂

File: mlir/include/mlir/Dialect/LLVMIR/LLVMAttrDefs.td
Comment (Radu2k) [reply mlir/include/mlir/Dialect/LLVMIR/LLVMAttrDefs.td:56]: I do not think this is the issue, since I have tried this and it only adds angle brackets.
`llvm.func @frame_pointer_func() vscale_range(1, 2) attributes {frame_pointer = #llvm<<"non-leaf">>} {`

Comment (joker-eph) [reply mlir/include/mlir/Dialect/LLVMIR/LLVMAttrDefs.td:56]: You didn't provide the mnemonic?

What about `def FramePointerKindAttr : LLVM_Attr<"FramePointerKind", "fp"> {`?

Comment (Radu2k) [reply mlir/include/mlir/Dialect/LLVMIR/LLVMAttrDefs.td:56]: Thanks for the suggestion, @joker-eph!

As mentioned above, even if the mnemonic is set to either 'fp' or 'framePointerKind', the results is still `#llvm.[mnemonic]<[value]> `, repectively (#llvm.fp<"non-leaf">, #llvm.framePointerKind<"non-leaf">). 

Setting the mnemonic to `framePointerKind` and adding `'<' '>'` makes lowering from MLIR to LLVM IR to work appropriately but in MLIR the parameter has to be assigned as `llvm.func @frame_pointer_func() attributes {frame_pointer = #llvm.framePointerKind<"non-leaf">} `. 

is this the right behaviour? If so, I will go ahead and update the patch.

Comment (Dinistro) [reply mlir/include/mlir/Dialect/LLVMIR/LLVMAttrDefs.td:56]: This matches what we want, I think.

Comment (joker-eph) [reply mlir/include/mlir/Dialect/LLVMIR/LLVMAttrDefs.td:56]: I would rather not mix inherent attribute with discardable attributes. But that's not gonna be automatic: you need to modify LLVMFuncOp::print() / LLVMFuncOp::parse() implementations.

Comment (Dinistro) [reply mlir/include/mlir/Dialect/LLVMIR/LLVMAttrDefs.td:56]: @joker-eph this is what was originally implemented.
@gysit and I previously suggested to not do that, to avoid bloating the printed form further. There are already inherent attributes that are mixed with the discardable attributes.
I suggest to ship this PR with the current approach and decide if we should add special parsing and printing for all inherent attributes or not after that.

Comment (joker-eph) [reply mlir/include/mlir/Dialect/LLVMIR/LLVMAttrDefs.td:56]: I don’t quite understand how pulling this before the { would be more bloated than after?

Comment (gysit) [reply mlir/include/mlir/Dialect/LLVMIR/LLVMAttrDefs.td:56]: The advantage of printing things in the attribute dictionary is that all attributes print the same way: "attribute_name = attribute_value". If it is printed outside, the function to print and parse is implemented by hand and it seems a bit harder to ensure consistency / problems with ordering during parsing etc.

If we want to clearly separate, discardable and inherent attributes we may also print it outside of the dictionary. I that case I would suggest to follow the style of vscale/comdat etc, i.e., `frame_pointer(non-leaf)` or similar

Comment (Radu2k) [reply mlir/include/mlir/Dialect/LLVMIR/LLVMAttrDefs.td:56]: In the end, should we proceed with the latest implementation or the first submission? I understand that there are certain advantages and disadvantages but which one would be the one desired, just to confirm.

Comment (gysit) [reply mlir/include/mlir/Dialect/LLVMIR/LLVMAttrDefs.td:56]: I would suggest to proceed with the latest implementation, but let's wait to see if @joker-eph agrees.

**Comment by Radu2k** [comment]
Hi, since some days have passed, I have updated the code to match the latest requirements. @joker-eph Please review the changes and let me know if these changes are alright with you. Many thanks to you all and if there is anything needed to be changed, please let me know.

**Review by gysit — COMMENTED** [review]
Can you rebase the revision? I think we should go ahead with the current implementation!

File: mlir/test/Target/LLVMIR/Import/frame-pointer.ll
Comment (Radu2k) [reply mlir/test/Target/LLVMIR/Import/frame-pointer.ll:5]: After rebasing, there is an updated format on the output when importing from llvm IR. 

`llvm.func @frame_pointer_func() attributes {frame_pointer = #llvm.framePointerKind<"non-leaf">, passthrough = [["frame-pointer", "non-leaf"]]} {`

There is this new 'passthrough' parameter. Is this alright? For me it his seems to be duplicated information and I am not sure if there is any way to specify if this should be outputted or not.

**Comment by Dinistro** [comment]
Please do not close a PR with the discussion, just rebase and force push to the same branch.

File: mlir/test/Target/LLVMIR/Import/frame-pointer.ll
Comment (Dinistro) [inline mlir/test/Target/LLVMIR/Import/frame-pointer.ll:7]:
```suggestion
```

**Review by Dinistro — APPROVED** [review]
LGTM modulo ultra nit.
Let me know when I should land this for you.

**Comment by Radu2k** [comment]
@Dinistro I have already started working on the pass implementation so, as soon as this has landed will carry on with submitting the pass PR. Please let me know if we should add you, as well as @gysit, as reviewers there as well or you have any suggestions for reviewers. Thx!

**Comment by Dinistro** [comment]
Feel free to add us as reviewers :)

**Pull request merged**
